### PR TITLE
Implement TLV in CASE Sigma Protocol

### LINKED ISSUE #5934: CASE handshake implementation looks nothing like the spec
#### Problem
#5442 implemented CASE, but it looks very little like the CASE in the spec in terms of what's going in to the messages and how.

 #### Proposed Solution
Implement what the spec says.

### PR DESCRIPTION
**Pull request by jpk233**
#### Problem
What is being fixed?  Examples:
* CASE does not currently follow the current spec with message handling.
* Fixes #5934

#### Change overview
* Replaced Sigma Message Construction using BoundBuf with TLV

#### Testing
How was this tested? (at least one bullet point required)
* Tested using CASE Session Unit Tests

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]
@pan-apple ?

**Review by jmartinez-silabs — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.cpp [hunk 1]
```diff
@@ -358,47 +393,57 @@ CHIP_ERROR CASESession::SendSigmaR1()
...
-    err = FindValidTrustedRoot(&buf, n_trusted_roots);
+    err = FindValidTrustedRoot(tlvReader, n_trusted_roots);
     SuccessOrExit(err);
+
     // write public key from message
-    bbuf.Put(buf, kP256_PublicKey_Length);
-    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+    err = tlvReader.FindElementWithTag(CASETLVTag::kInitiatorEphPubKey, suppTlvReader);
+    SuccessOrExit(err);
+    VerifyOrExit(mRemotePubKey.Length() == suppTlvReader.GetLength(), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
```
Comment (emargolis) [inline src/protocols/secure_channel/CASESession.cpp:442]:
```suggestion
    VerifyOrExit(mRemotePubKey.Length() == suppTlvReader.GetLength(), err = CHIP_ERROR_INVALID_TLV_ELEMENT);
```

File: src/protocols/secure_channel/CASESession.cpp [hunk 2]
```diff
@@ -479,67 +524,82 @@ CHIP_ERROR CASESession::SendSigmaR2()
...
+    {
+        System::PacketBufferTLVWriter tlvWriter;
+        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+        tlvWriter.Init(std::move(msg_R2_Encrypted));
+        SuccessOrExit(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kNOC, mOpCredSet->GetDevOpCred(mTrustedRootId),
+                                               mOpCredSet->GetDevOpCredLen(mTrustedRootId)));
+        SuccessOrExit(
+            err = tlvWriter.PutBytes(CASETLVTag::kSignature, sigmaR2Signature, static_cast<uint32_t>(sigmaR2Signature.Length())));
```
Comment (emargolis) [inline src/protocols/secure_channel/CASESession.cpp:567]: According to the spec: Section `4.12.2.5. Signature` the signature should be encoded in its raw 64 representation over the wire.

File: src/protocols/secure_channel/CASESession.cpp [hunk 3]
```diff
@@ -479,67 +524,82 @@ CHIP_ERROR CASESession::SendSigmaR2()
...
+    err = AES_CCM_encrypt(msg_R2_Encrypted->Start(), msg_R2_Encrypted->DataLength(), nullptr, 0, sr2k, kAEADKeySize, kIVSR2,
+                          kIVLength, msg_R2_Encrypted->Start(), tag, sizeof(tag));
     SuccessOrExit(err);
 
     data_len = static_cast<uint16_t>(kSigmaParamRandomNumberSize + sizeof(uint16_t) + kTrustedRootIdSize + kP256_PublicKey_Length +
-                                     msg_r2_signed_enc_len + sizeof(tag));
+                                     msg_R2_Encrypted->DataLength() + sizeof(tag));
 
-    msg_R2 = System::PacketBufferHandle::New(data_len);
+    msg_R2 = System::PacketBufferHandle::New(data_len + System::PacketBuffer::kDefaultHeaderReserve);
```
Comment (emargolis) [inline src/protocols/secure_channel/CASESession.cpp:580]: Isn't `System::PacketBuffer::kDefaultHeaderReserve` automatically added by default when New() is called?

Similar comment for other placed in this PR.

File: src/protocols/secure_channel/CASESession.cpp [hunk 4]
```diff
@@ -639,47 +710,81 @@ CHIP_ERROR CASESession::HandleSigmaR2(const System::PacketBufferHandle & msg)
...
-        MutableByteSpan msg_R2_Span(msg_R2_Signed.Get(), msg_r2_signed_len);
-        err = ConstructSignedCredentials(&buf, remoteDeviceOpCert, remoteDeviceOpCertLen, msg_R2_Span, sigmaR2SignedData, sigLen);
-        SuccessOrExit(err);
-    }
+    err = ConstructTBS2Data(responderOpCert, responderOpCertLen, msg_R2_Signed);
+    SuccessOrExit(err);
+
+    err = tlvReader.FindElementWithTag(CASETLVTag::kSignature, suppTlvReader);
+    SuccessOrExit(err);
+    VerifyOrExit(suppTlvReader.GetType() == TLV::kTLVType_UTF8String || suppTlvReader.GetType() == TLV::kTLVType_ByteString,
```
Comment (emargolis) [inline src/protocols/secure_channel/CASESession.cpp:780]:
```suggestion
    VerifyOrExit(suppTlvReader.GetType() == TLV::kTLVType_ByteString,
```

According to the spec the signature SHALL be encoded as OCTET STRING. Why UTF8String is also defined as an option here? 

Similar comment applies to many other placed in this PR.

File: src/protocols/secure_channel/CASESession.cpp [hunk 5]
```diff
@@ -69,6 +68,39 @@ using HKDF_sha_crypto = HKDF_sha;
 // The session establishment fails if the response is not received within timeout window.
 static constexpr ExchangeContext::Timeout kSigma_Response_Timeout = 10000;
 
+/**
+ * \brief
+ * A list of registered packet types a.k.a. TLV context-specific tags to be used during CASE protocol.
+ */
+enum CASETLVTag : uint64_t
```
[the remark this answers is not in the document]
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp]: Changed to uint8.

File: src/protocols/secure_channel/CASESession.cpp (hunk 2 above)
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp:568]: We will address this in a separate commit where we'll bring the protocol up to the current spec. The intention for this specific PR is just to add TLV format.

File: src/protocols/secure_channel/CASESession.cpp (hunk 3 above)
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp]: Updated.

File: src/protocols/secure_channel/CASESession.cpp (hunk 4 above)
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp]: Updated.

**Review by emargolis — APPROVED** [review]

**Review by emargolis — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.cpp [hunk 6]
```diff
@@ -562,40 +623,43 @@ CHIP_ERROR CASESession::SendSigmaR2()
...
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
 
     const uint8_t * buf = msg->Start();
     size_t buflen       = msg->DataLength();
 
     chip::Platform::ScopedMemoryBuffer<uint8_t> msg_salt;
     uint16_t saltlen;
 
-    chip::Platform::ScopedMemoryBuffer<uint8_t> msg_R2_Signed;
+    System::PacketBufferHandle msg_R2_Encrypted;
```
Comment (pan-apple) [inline src/protocols/secure_channel/CASESession.cpp:647]: Let's not use PacketBufferHandle for internal buffers that are not actually sent. ScopedMemoryBuffers are better suited for such allocations.

The system could have limited PacketBufferHandle pool and we should use it only when necessary.

This comment applies to multiple areas in this file.

**Review by pan-apple — CHANGES_REQUESTED** [review]
@jpk233, let us not use PacketBufferHandle where not needed. It puts extra burden on the pool. In my testing I had noticed the pool was getting depleted on repeated tests. Even though the buffer is not leaked, but it can temporarily deplete the pool enough to cause issues.

File: src/protocols/secure_channel/CASESession.cpp (hunk 6 above)
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp]: Replaced internal buffers using PacketBufferHandle with ScopedMemoryBuffers.

File: src/protocols/secure_channel/CASESession.cpp (hunk 1 above)
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp]: Updated.

File: src/protocols/secure_channel/CASESession.cpp [hunk 7]
```diff
@@ -70,9 +69,42 @@ using HKDF_sha_crypto = HKDF_sha;
...
+    kTag = 10,
+    // TODO: Remove tag 11
+    /*! \brief Tag 11. The packet contains the total number of Trusted Root IDs. */
+    kNumberofTrustedRootIDs = 11,
+};
+
 CASESession::CASESession()
 {
-    mTrustedRootId = CertificateKeyId();
+    mTrustedRootId.mId = nullptr;
```
[the remark this answers is not in the document]
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp]: Updated to no longer conflict.

File: src/protocols/secure_channel/CASESession.cpp [hunk 8]
```diff
@@ -359,47 +394,56 @@ CHIP_ERROR CASESession::SendSigmaR1()
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR CASESession::HandleSigmaR1_and_SendSigmaR2(const System::PacketBufferHandle & msg)
+CHIP_ERROR CASESession::HandleSigmaR1_and_SendSigmaR2(System::PacketBufferHandle & msg)
```
Comment (pan-apple) [inline src/protocols/secure_channel/CASESession.cpp:397]: Wondering why the `const` was removed (here and at other message handlers)?

**Comment by pan-apple** [comment]
> @jpk233, let us not use PacketBufferHandle where not needed. It puts extra burden on the pool. In my testing I had noticed the pool was getting depleted on repeated tests. Even though the buffer is not leaked, but it can temporarily deplete the pool enough to cause issues.

Thanks for resolving this comment!

File: src/protocols/secure_channel/CASESession.cpp (hunk 8 above)
Comment (jpk233) [reply src/protocols/secure_channel/CASESession.cpp:397]: This is because PacketBufferTLVReader Class' Init method requires a non-const PacketBufferHandle

**Review by pan-apple — APPROVED** [review]

**Review by emargolis — APPROVED** [review]

**Review by mspang — APPROVED** [review]
rubber stamping @emargolis's review

**Pull request merged**